CMakeLists.txt

```diff
@@ -29,36 +28,136 @@ set( CMAKE_ARCHIVE_OUTPUT_DIRECTORY
 list( APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake" )
 include( traccc-compiler-options )
 
-# option for algebra plugins (ARRAY EIGEN SMATRIX VC VECMEM)
-set(TRACCC_ALGEBRA_PLUGIN EIGEN)
-if (TRACCC_ALGEBRA_PLUGIN)
-  message(STATUS "Building with plugin type: " ${TRACCC_ALGEBRA_PLUGIN})
-  set(ALGEBRA_PLUGIN_INCLUDE_${TRACCC_ALGEBRA_PLUGIN} ON)
-  add_definitions(-DALGEBRA_PLUGIN_INCLUDE_${TRACCC_ALGEBRA_PLUGIN})
+# Check if CUDA is available.
+include( CheckLanguage )
+check_language( CUDA )
+set( TRACCC_BUILD_CUDA_DEFAULT FALSE )
+if( CMAKE_CUDA_COMPILER )
+   set( TRACCC_BUILD_CUDA_DEFAULT TRUE )
 endif()
 
-# option for algebra scalar type 
-set(TRACCC_CUSTOM_SCALARTYPE float)
-if(TRACCC_CUSTOM_SCALARTYPE)
-    message(STATUS "Building with custom scalar type: " ${TRACCC_CUSTOM_SCALARTYPE})
-    set(ALGEBRA_PLUGIN_CUSTOM_SCALARTYPE ${TRACCC_CUSTOM_SCALARTYPE})
+# Flags controlling which parts of traccc to build.
+option( TRACCC_BUILD_CUDA "Build the CUDA sources included in traccc"
+   ${TRACCC_BUILD_CUDA_DEFAULT} )
+option( TRACCC_BUILD_TESTING "Build the (unit) tests of traccc" TRUE )
+
+# Clean up.
+unset( TRACCC_BUILD_CUDA_DEFAULT )
+
+# Set up VecMem.
+option( TRACCC_SETUP_VECMEM
+   "Set up the VecMem target(s) explicitly" TRUE )
```

Comment:
Does it make sense to have this be a configurable parameter? Surely if you turn this off there is no way your code will ever compile, right?

Reply:
Are you absolutely sure that traccc will only ever be a top-level project? We have the same sort of flag for both algebra-plugins and detray. Because I knew that I will need to turn both of those off once we build them as part of this project.

The flag is here to allow us to potentially use this as a sub-project of even higher level projects.

Reply:
That's not really what I mean - if you turn this off then CMake will never find vecmem at all, right? So then what will it know what to link against?

Reply:
If this project is added with [add_subdirectory](https://cmake.org/cmake/help/latest/command/add_subdirectory.html) to some parent project, then that parent project could've set up the build of VecMem itself. At which point either using `find_package(...)` or trying to set up the VecMem build itself, would be a bad choice in this place.

If we want to allow the usage of this project as a sub-project of something even higher level (like the repository of one of you guys that you use to make measurements for your thesis, against a fixed version of this repository), then it must be possible to tell it not to set up any of its externals. And just expect "the higher level project" to set those up outside of this code.

Reply:
Hold on are you saying that we are inheriting dependencies from the parent project??? :fearful: